ENH: Add script to write the event files

esavary opened the pull request:
- Add script to extract the event files from the output of phys2BIDS (compressed tsv files).
- The script saves the event file in the same folder as the compressed tsv files.
- The script generates PNG plots for each task, displaying both physiological data and corresponding events.

closes #90

esavary committed:
Add write_event_file.py

esavary committed:
ENH: add heuristic file

- Add heuristic file for phys2bids.
- Add commented code for the future version of BHT.

esavary committed:
ENH: Add writing the events json file

esavary committed:
ENH: improve json event file

esavary committed:
ENH: Add function to plot physiological data

- Add function to plot physiological data
- Fix the naming of the files

esavary committed:
ENH: Process all files at once

- Add a function to write the event files for all physiological data files of a folder.
- Take the path of the folder as a command-line argument.

esavary committed:
FIX: fix heur_physio.py

- Add a condition to process all .acq files.
- remove info['rec'].

esavary committed:
ENH: Add instructions to process physiological data

esavary committed:
ENH: Add precisions on how to process physiological data

esavary committed:
FIX: correct event duration

esavary commented:
The task durations are currently hardcoded in the write_event_file.py script. These values are also hardcoded in the Python files generated by Psychopy. If we continue using the .psyexp files, it might be a more straightforward and cleaner approach to extract and parse the task durations directly from there instead of from the Python files. What are your thoughts on this @oesteban ?

oesteban commented on code/physioconv/write_event_file.py:
```diff
@@ -0,0 +1,263 @@
+            """
+        json_content = {
+            "trial_type": {
+                "LongName": "Event category",
```
```suggestion
                "LongName": "Breath-holding task conditions (that is, breath-in, breath-out, and hold)",
```

oesteban commented on code/physioconv/write_event_file.py:
```diff
@@ -0,0 +1,263 @@
+                    "SoftwareName": "PsychoPy",
+                    "SoftwareRRID": "SCR_006571",
+                    "SoftwareVersion": "2022.3.0.dev6",
+                    "Code": "http://www.psychopy.org",
```
```suggestion
                    "Code": "https://github.com/TheAxonLab/HCPh-fMRI-tasks/blob/97cc7879622f45129eefb9968890b41631f40851/task-bht_bold.psyexp",
```

oesteban commented on code/physioconv/write_event_file.py:
```diff
@@ -0,0 +1,263 @@
+                hold = {"onset": row[0], "duration": 2.7, "trial-type": "hold"}
+                event_dataframe = event_dataframe.append(hold, ignore_index=True)
+            """
+        json_content = {
```
Better to avoid code repetitions. Initialize the json file above and then you just copy the dictionary here

oesteban replied:
For each task:

```suggestion
        json_content = EVENTS_JSON_BOILERPLATE.copy()
        json_contents["StimulusPresentation"]["Code"] = "https://github.com/TheAxonLab/HCPh-fMRI-tasks/blob/97cc7879622f45129eefb9968890b41631f40851/task-bht_bold.psyexp"
        json_contents["Description"] = "Indicator of type of action that is expected"
        json_contents["LongName"] = "Breath-holding task conditions (that is, breath-in, breath-out, and hold)"
        json_contents["Levels"] = {
            "breath-in": "A green rectangle is displayed to indicate breathing in",
            "breath-out": "A yellow rectangle (orange for the last breath-in before hold) is displayed to indicate breathing out",
            "hold": "A red rectangle is displayed to indicate breath hold",
        }
```

oesteban commented on code/physioconv/write_event_file.py:
```diff
@@ -0,0 +1,263 @@
+import matplotlib.pyplot as plt
+import matplotlib.patches as mpatches
+
+
```
```suggestion
EVENTS_JSON_BOILERPLATE = {
    "StimulusPresentation": {
        "OperatingSystem": "Linux Ubuntu 20.04.5",
        "SoftwareName": "PsychoPy",
        "SoftwareRRID": "SCR_006571",
        "SoftwareVersion": "2022.3.0.dev6"
    }
}

```

esavary committed:
Update code/physioconv/write_event_file.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update code/physioconv/write_event_file.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update code/physioconv/write_event_file.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update code/physioconv/write_event_file.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
FIX: modified json file code

- modified json file structure for bht to match https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/05-task-events.html.
- modified json file code for pct and rs to match the code style of bht.

oesteban commented on code/physioconv/write_event_file.py:
```diff
@@ -0,0 +1,247 @@
+    """
+    Plot physiological data along with events.
+
+    Args:
+        time_series_df (pd.DataFrame): DataFrame containing the physiological data.
+        events_df (pd.DataFrame): DataFrame containing the event data.
+        tsv_file (str): Path to the TSV file.
+        output_folder (str, optional): Output folder for saving the plot image. Default is current directory.
+
+    Returns:
+        None: The plot is saved as an PNG file.
```
```suggestion
    Parameters
    ----------
    time_series_df : :obj:`pandas.DataFrame`
        Table containing the physiological data.
    events_df : :obj:`pandas.DataFrame`
        Table containing the event data.
    tsv_file : :obj:`os.pathlike`
        Path to the TSV file.
    output_folder : :obj:`os.pathfile`
        Output folder for saving the plot image. Default is current directory.

    Returns
    -------
    The path where the plot was saved as a PNG file.
```

oesteban commented on code/physioconv/write_event_file.py:
```diff
@@ -0,0 +1,247 @@
+    )
+    plt.tight_layout()
+    plt.savefig(output_file)
+
```
```suggestion
    return output_file

```

esavary replied:
I accepted the suggestion, but what is the idea behind having only this function return a value?

oesteban replied:
What's the point of returning `None`?

I would honestly return the matplotlib figure object and save the figure outside the function, but returning the filename seemed to me like the least impactful change giving some extra value to the function.

oesteban commented on code/physioconv/write_event_file.py:
```diff
@@ -0,0 +1,247 @@
+    """
+    This function reads the trigger channels of the physiological file, processes them to create the event file.
+
+    Parameters:
+        tsv_file (str): The path to the input gzipped TSV file containing physiological data of a task.
```
```suggestion
    Parameters
    ----------
    tsv_file :obj:`os.pathlike`
         The path to the input gzipped TSV file containing physiological data of a task.
```

oesteban commented on code/physioconv/write_event_file.py:
```diff
@@ -0,0 +1,247 @@
+    """
+    Find all files in the given folder with names containing "_physio.tsv.gz", write the corresponding event files and save a plot of the physiological data.
+
+    Args:
+        folder_path (str): Path to the folder containing the files.
```
```suggestion
    Parameters
    ----------
    folder_path : :obj:`os.pathlike`
        Path to the folder containing the files.
```

oesteban commented on docs/data-management/post-session.md:
````diff
@@ -332,9 +332,38 @@ The output of *HeuDiConv* with our current heuristics and reproin conventions sh
 ```
 phys2bids -in /data/datasets/hcph-pilot/sourcedata/physio/session-recording.acq -info
 ```
+- [ ] Check the data by opening the PNG plot.
````
Unclear what is to be checked.

oesteban commented on docs/data-management/post-session.md:
````diff
@@ -332,9 +332,38 @@ The output of *HeuDiConv* with our current heuristics and reproin conventions sh
 phys2bids -in /data/datasets/hcph-pilot/sourcedata/physio/session-recording.acq -info
 ```
+- [ ] Check the data by opening the PNG plot.
+- [ ] If the data appears satisfactory, proceed to process it using the subsequent command. Use the -ntp argument to specify the number of volumes for each task, and the -tr argument to indicate the task's repetition time. Define the output directory with -outdir and provide the path to the heuristic file using -heur. Adjust the subject and session numbers accordingly. Should scanner trigger tran ...
````
Relatedly, what is "satisfactory" in this context?

```suggestion
- [ ] If the data appears satisfactory, proceed to process it using the subsequent command.
    Use the `-ntp` argument to specify the number of volumes for each task, and the `-tr` argument to indicate the task's repetition time.
    Define the output directory with `-outdir` and provide the path to the heuristic file using `-heur`.
    Adjust the subject and session numbers accordingly.
    Should scanner trigger transmission encounter issues and manual adjustments are made to the trigger data, it is possible to allocate one trigger per task.
    Set the repetition time duration as the task length, as demonstrated in the example below.
```

oesteban commented on docs/data-management/post-session.md:
```diff
@@ -332,9 +332,38 @@ The output of *HeuDiConv* with our current heuristics and reproin conventions sh
 
 ### Convert physiological recordings and eye-tracking data to BIDS
 
+- [ ] Plot an overview of the data with the following command. This command generates a PNG plot of the data within the current directory without processing the data itself. The folder containing the physiological data is specified via the -in command line argument.
```
Please split sentences in new lines; it's easier to review later and see changes.
```suggestion
- [ ] Plot an overview of the data with the following command.
     This command generates a PNG plot of the data within the current directory without processing the data itself.
     The physiological data folder is specified via the `-in` command line argument.
```

oesteban commented on docs/data-management/post-session.md:
````diff
@@ -332,9 +332,38 @@ The output of *HeuDiConv* with our current heuristics and reproin conventions sh
+```
+phys2bids -in modified-last-session_multiscan.txt -chtrig 4 -ntp  1 1 1 -tr 158 1200 331 -thr 2 -outdir outputdir -heur heur_physio.py -sub pilot -ses 01
+```
+- [ ] Execute the script write_event_file.py as shown below to generate task event files. This script not only creates JSON and TSV files containing event information, but it also generates PNG plots for each task, displaying both physiological data and corresponding events. These plots are saved in the current directory. The script must be executed with the following command, where outputdir is  ...
````
```suggestion
- [ ] Execute the script `write_event_file.py` as shown below to generate task event files.
    This script creates JSON and TSV files containing event information and generates PNG plots for each task, displaying both physiological data and corresponding events.
    These plots are saved in the current directory.
    The script must be executed with the following command, where `outputdir` is the output directory of *phys2bids*:
```

esavary committed:
Update code/physioconv/write_event_file.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update code/physioconv/write_event_file.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update code/physioconv/write_event_file.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update code/physioconv/write_event_file.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update docs/data-management/post-session.md

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update docs/data-management/post-session.md

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update docs/data-management/post-session.md

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Fix: correct typing

oesteban commented on code/physioconv/heur_physio.py:
````diff
@@ -0,0 +1,77 @@
+
+    Notes
+    -----
+    The `if ..` structure should always be similar to
+    ```
+    if physinfo == 'somepattern':
+        info['var'] = 'somethingelse'
+    ```
+    or, in case it's a partial match
+    ```
+    if fnmatch.fnmatchcase(physinfo, '*somepattern?'):
+        info['var'] = 'somethingelse'
+    ```
+    Where:
+        - `physinfo` and `info` are dedicated keywords,
+        - 'somepattern' is the name of the file,
+        - 'var' is a bids key in the list below
+        - 'somethingelse' is the value of the key
````
```suggestion
    The `if ..` structure should always be similar to::

        if physinfo == 'somepattern':
            info['var'] = 'somethingelse'

    or, in case it's a partial match::

        if fnmatch.fnmatchcase(physinfo, '*somepattern?'):
            info['var'] = 'somethingelse'

    where:
    
    - ``physinfo`` and ``info`` are dedicated keywords,
    - ``'somepattern'`` is the name of the file,
    - ``'var'`` is a BIDS key in the list below
    - ``'somethingelse'`` is the value of the key

```

oesteban commented on code/physioconv/heur_physio.py:
```diff
@@ -0,0 +1,77 @@
+    # ##  See example below          ## #
+    # ################################# #
+
+    if fnmatch.fnmatchcase(physinfo, "*multiscan*"):
+        if take == "01":
+            info["task"] = "qct"
+        if take == "02":
+            info["task"] = "rest"
+        if take == "03":
+            info["task"] = "bht"
```
This is not for us, I guess.
```suggestion
```

oesteban commented on code/physioconv/heur_physio.py:
```diff
@@ -0,0 +1,77 @@
+            info["task"] = "rest"
+        if take == "03":
+            info["task"] = "bht"
+    elif fnmatch.fnmatchcase(physinfo, "*.acq*"):
+        if take == "01":
+            info["task"] = "qct"
+        if take == "02":
+            info["task"] = "rest"
+        if take == "03":
+            info["task"] = "bht"
```
```suggestion
    if not fnmatch.fnmatchcase(physinfo, "*.acq*"):
        raise RuntimeError("Unknown input file")
    
    TAKES = {
        "01": "qct",
        "02": "rest",
        "03": "bht",
    }
    info["task"] = TAKES[take]
```

Question: what happens when we have failed runs?

oesteban commented:
> The task durations are currently hardcoded in the write_event_file.py script. These values are also hardcoded in the Python files generated by Psychopy. If we continue using the .psyexp files, it might be a more straightforward and cleaner approach to extract and parse the task durations directly from there instead of from the Python files. What are your thoughts on this @oesteban ?

Let's go with a JSON or a YAML file to store this.

oesteban commented:
Perhaps YAML to allow multiline strings (I can imagine we will end up writing descriptions and things like that in this file)

esavary committed:
Fix: add details on checking physiological data

oesteban committed:
sty: fix long strings, run black, edit docstrings

oesteban committed:
Merge branch 'physio-to_BIDS' of github.com:esavary/hcph-sops into physio-to_BIDS

esavary committed:
Fix: fix typo

oesteban commented:
we should squash and merge instead of a direct merge of this one (27 commits and counting...)

oesteban committed:
Apply suggestions from code review
